### PR TITLE
[ibm think videos] Added new extractor

### PR DESCRIPTION
**Pull request by parth-verma**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Adds the ability to download videos and playlist from IBM think videos thus closing #15971

### CODE REVIEW

File: youtube_dl/extractor/ibm_think.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            'uploader': 'f8k4md3yana',
+            'uploader_id': '43178333',
+        }
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        ustream_url = self._html_search_regex(r'<iframe\ssrc="(.*?)"', webpage, 'ustream_url').split('?')[0] + '/'
+        return self.url_result(ustream_url, UstreamIE.ie_key())
```
Comment (dstftw) [inline youtube_dl/extractor/ibm_think.py]: Must be handled by generic extractor.

File: youtube_dl/extractor/ibm_think.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            'description': 'Discover what the world is thinking at Think 2018, IBM\'s first business event to go beyond IT conference, exploring cloud technology, data analytics & security.'
+        },
+        'playlist_mincount': 6
+    }]
+
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+        webpage = self._download_webpage(url, playlist_id)
+        entries = [self.url_result(m) for m in re.findall(r'<a href="(.+?)" class="(?:video-list-item)?\s?(?:js-video-list-item)?">', webpage)]
+        title = self._html_search_regex(r'<title>(?:.+?)\s\|\s(?:.+?)\s\|\s(.+?)</title>', webpage, 'title')
```
Comment (dstftw) [inline youtube_dl/extractor/ibm_think.py]: Invalid regex. Did you even run this code? Should not be fatal.

File: youtube_dl/extractor/ibm_think.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            'title': 'Five innovations that will help change our lives within five years',
+            'description': 'Discover what the world is thinking at Think 2018, IBM\'s first business event to go beyond IT conference, exploring cloud technology, data analytics & security.'
+        },
+        'playlist_mincount': 6
+    }]
+
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+        webpage = self._download_webpage(url, playlist_id)
+        entries = [self.url_result(m) for m in re.findall(r'<a href="(.+?)" class="(?:video-list-item)?\s?(?:js-video-list-item)?">', webpage)]
```
Comment (dstftw) [inline youtube_dl/extractor/ibm_think.py]: Invalid regex.

**Comment by party98** [comment]
@dstftw As requested I have made the changes. Please review for any more possible issues (if any).

File: youtube_dl/extractor/ibm_think.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            'uploader': 'f8k4md3yana',
+            'uploader_id': '43178333',
+        }
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        ustream_url = self._html_search_regex(r'<iframe\ssrc="(.*?)"', webpage, 'ustream_url').split('?')[0] + '/'
+        return self.url_result(ustream_url, GenericIE.ie_key())
```
Comment (dstftw) [inline youtube_dl/extractor/ibm_think.py]: Nothing changed.

File: youtube_dl/extractor/ibm_think.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            'description': 'Discover what the world is thinking at Think 2018, IBM\'s first business event to go beyond IT conference, exploring cloud technology, data analytics & security.'
+        },
+        'playlist_mincount': 6
+    }]
+
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+        webpage = self._download_webpage(url, playlist_id)
+        entries = [self.url_result(m, GenericIE.ie_key()) for m in re.findall(r'<a href="(.+?)" class="(?:video-list-item)?\s?(?:js-video-list-item)?">', webpage)]
+        title = self._html_search_regex(r'<title>(?:.+?)\s\|\s(?:.+?)\s\|\s(.+?)</title>', webpage, 'title')
```
Comment (dstftw) [inline youtube_dl/extractor/ibm_think.py]: Nothing changed.

File: youtube_dl/extractor/ibm_think.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            'title': 'Five innovations that will help change our lives within five years',
+            'description': 'Discover what the world is thinking at Think 2018, IBM\'s first business event to go beyond IT conference, exploring cloud technology, data analytics & security.'
+        },
+        'playlist_mincount': 6
+    }]
+
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+        webpage = self._download_webpage(url, playlist_id)
+        entries = [self.url_result(m, GenericIE.ie_key()) for m in re.findall(r'<a href="(.+?)" class="(?:video-list-item)?\s?(?:js-video-list-item)?">', webpage)]
```
Comment (dstftw) [inline youtube_dl/extractor/ibm_think.py]: Nothing changed.

File: youtube_dl/extractor/ibm_think.py (hunk 4 above)
Comment (party98) [reply youtube_dl/extractor/ibm_think.py]: I am using the generic ie here as requested.

File: youtube_dl/extractor/ibm_think.py (hunk 5 above)
Comment (party98) [reply youtube_dl/extractor/ibm_think.py]: The regex here is valid for all the tests and i was able to successfully download the video from the links mentioned in the issue and some other as well

File: youtube_dl/extractor/ibm_think.py (hunk 4 above)
Comment (dstftw) [reply youtube_dl/extractor/ibm_think.py]: This was address to complete extractor.

File: youtube_dl/extractor/ibm_think.py (hunk 6 above)
Comment (party98) [reply youtube_dl/extractor/ibm_think.py]: The regex here is valid for all the tests and i was able to successfully download the video from the links mentioned in the issue and some other as well.

File: youtube_dl/extractor/ibm_think.py (hunk 4 above)
Comment (party98) [reply youtube_dl/extractor/ibm_think.py]: I am unable to get the issue. Please explain in a little more detail as i am new to this project.

File: youtube_dl/extractor/ibm_think.py (hunk 5 above)
Comment (dstftw) [reply youtube_dl/extractor/ibm_think.py]: Non capturing groups are superfluous.
Playlist title must not be fatal, read coding conventions.

File: youtube_dl/extractor/ibm_think.py (hunk 6 above)
Comment (dstftw) [reply youtube_dl/extractor/ibm_think.py]: This regex will match any href with empty `class` that does not make any sense.

File: youtube_dl/extractor/ibm_think.py (hunk 4 above)
Comment (dstftw) [reply youtube_dl/extractor/ibm_think.py]: Generic iframe embeds must be extracted by `GenericIE`. Do not add new extractors for this.

**Comment by party98** [comment]
@dstftw Please review now.

File: youtube_dl/extractor/ibm_think.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+            'title': 'Five innovations that will help change our lives within five years',
+            'description': 'Discover what the world is thinking at Think 2018, IBM\'s first business event to go beyond IT conference, exploring cloud technology, data analytics & security.'
+        },
+        'playlist_mincount': 6
+    }]
+
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+        webpage = self._download_webpage(url, playlist_id)
+        entries = [self.url_result(m, GenericIE.ie_key()) for m in re.findall(r'<a href="(.+?)" class="video-list-item js-video-list-item?">', webpage)]
```
Comment (dstftw) [inline youtube_dl/extractor/ibm_think.py]: >item?
>m?

File: youtube_dl/extractor/ustream.py [hunk 8]
```diff
@@ -72,7 +72,7 @@ class UstreamIE(InfoExtractor):
     @staticmethod
     def _extract_url(webpage):
         mobj = re.search(
-            r'<iframe[^>]+?src=(["\'])(?P<url>http://www\.ustream\.tv/embed/.+?)\1', webpage)
+            r'<iframe[^>]+?src=(["\'])(?P<url>https?://www\.ustream\.tv/embed/.+?)\1', webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/ustream.py]: Make `https?:` part optional.

Comment (party98) [reply youtube_dl/extractor/ustream.py]: so something like `www.ustream.tv/embed/adsda` should work?

Comment (dstftw) [reply youtube_dl/extractor/ustream.py]: No.

Comment (party98) [reply youtube_dl/extractor/ustream.py]: then what do you mean by making the `https:?` part optional as the `?` will make it compatible for both `http` and `https`

Comment (dstftw) [reply youtube_dl/extractor/ustream.py]: Re-read my post with eyes please.

Comment (party98) [reply youtube_dl/extractor/ustream.py]: sorry but I am still unable to get what the requirement is, can you please explain with an example.

Comment (kayb94) [reply youtube_dl/extractor/ustream.py]: the whole thing i guess, like e.g. in (from vimeo.py line 457):
r'<video[^>]+src=(["\'])(?P<url>(?:https?:)?//(?:www\.)?vimeo\.com/[0-9]+)\1'

**Pull request closed without merging**